{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a95d8671-5a40-4480-8896-cd9656fc2ef9",
   "metadata": {},
   "source": [
    "# Question 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23d50276-7117-4bcd-8537-5118fb98158c",
   "metadata": {},
   "source": [
    "## Imports and functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "13702700-c272-4f63-8e9a-609b665533dc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-06-30 18:43:09.669070: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-06-30 18:43:10.609766: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/intel/compilers_and_libraries_2018.1.163/linux/tbb/lib/intel64_lin/gcc4.7:/opt/intel/compilers_and_libraries_2018.1.163/linux/compiler/lib/intel64_lin:/opt/intel/compilers_and_libraries_2018.1.163/linux/mkl/lib/intel64_lin::/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64/\n",
      "2024-06-30 18:43:10.609914: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/intel/compilers_and_libraries_2018.1.163/linux/tbb/lib/intel64_lin/gcc4.7:/opt/intel/compilers_and_libraries_2018.1.163/linux/compiler/lib/intel64_lin:/opt/intel/compilers_and_libraries_2018.1.163/linux/mkl/lib/intel64_lin::/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64/\n",
      "2024-06-30 18:43:10.609927: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.\n"
     ]
    }
   ],
   "source": [
    "from sentence_transformers import SentenceTransformer\n",
    "import datasets\n",
    "from datasets import load_dataset, list_datasets\n",
    "from pinecone import Pinecone, ServerlessSpec\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "import cohere\n",
    "import numpy as np\n",
    "import warnings\n",
    "from IPython.display import display\n",
    "import pandas as pd\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fdb85cbf-644c-492c-8b3d-a9930dcc3a65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pipeline functions (based on tutorial 3)\n",
    "\n",
    "def load_and_embedd_dataset(\n",
    "        dataset_name: str = 'cnn_dailymail',\n",
    "        split: str = 'train',\n",
    "        model: SentenceTransformer = SentenceTransformer('all-MiniLM-L6-v2'),\n",
    "        text_field: str = 'highlights',\n",
    "        rec_num: int = 400,\n",
    "        shuffle: bool = False,\n",
    "        shuffle_seed: int = 3435,\n",
    "        semantic_chunk: bool = False,\n",
    "        chunk_window: int = 2,\n",
    "        chunk_overlap: int = 1,\n",
    "        chunk_percentile_threshold: int = 90,\n",
    "        preprocess_func = None,\n",
    ") -> tuple:\n",
    "    \"\"\"\n",
    "    Load a dataset and embedd the text field using a sentence-transformer model\n",
    "    Args:\n",
    "        dataset_name: The name of the dataset to load\n",
    "        split: The split of the dataset to load\n",
    "        model: The model to use for embedding\n",
    "        text_field: The field in the dataset that contains the text\n",
    "        rec_num: The number of records to load and embedd\n",
    "        shuffle: Wheter to shuffle the dataset\n",
    "        shuffle_seed: the seed for random shuffling (duh)\n",
    "        semantic_chunk: wheter to semantically chunk the documents before embedding\n",
    "    Returns:\n",
    "        tuple: A tuple containing the dataset and the embeddings\n",
    "    \"\"\"\n",
    "    from datasets import load_dataset\n",
    "    \n",
    "    print(\"Loading and embedding the dataset\")\n",
    "    \n",
    "    # Load the dataset\n",
    "    dataset = load_dataset(dataset_name, split=split)\n",
    "    # Preprocess (if needed)\n",
    "    if preprocess_func is not None:\n",
    "        dataset = dataset.map(preprocess_func)\n",
    "    # Shuffle (if needed)\n",
    "    if shuffle:\n",
    "        dataset = dataset.shuffle(seed=shuffle_seed)\n",
    "    # Semantic chunking (if needed)\n",
    "    if semantic_chunk:\n",
    "        dataset, rec_num = chunk_dataset(dataset, window=chunk_window, overlap=chunk_overlap, \n",
    "                                percentile_threshold=chunk_percentile_threshold, rec_num=rec_num)\n",
    "    \n",
    "    # Embed the first `rec_num` rows of the dataset  \n",
    "    embeddings = model.encode(dataset[text_field][:rec_num])\n",
    "    \n",
    "    print(\"Done!\")\n",
    "    return dataset, embeddings\n",
    "\n",
    "\n",
    "def create_pinecone_index(\n",
    "        index_name: str,\n",
    "        dimension: int,\n",
    "        metric: str = 'cosine',\n",
    ") -> Pinecone:\n",
    "    \"\"\"\n",
    "    Create a pinecone index if it does not exist\n",
    "    Args:\n",
    "        index_name: The name of the index\n",
    "        dimension: The dimension of the index\n",
    "        metric: The metric to use for the index\n",
    "    Returns:\n",
    "        Pinecone: A pinecone object which can later be used for upserting vectors and connecting to VectorDBs\n",
    "    \"\"\"\n",
    "    from pinecone import Pinecone, ServerlessSpec\n",
    "    print(\"Creating a Pinecone index...\")\n",
    "    pc = Pinecone(api_key=PINECONE_API_KEY)\n",
    "    existing_indexes = [index_info[\"name\"] for index_info in pc.list_indexes()]\n",
    "    if index_name not in existing_indexes:\n",
    "        pc.create_index(\n",
    "            name=index_name,\n",
    "            dimension=dimension,\n",
    "            # Remember! It is crucial that the metric you will use in your VectorDB will also be a metric your embedding\n",
    "            # model works well with!\n",
    "            metric=metric,\n",
    "            spec=ServerlessSpec(\n",
    "                cloud=\"aws\",\n",
    "                region=\"us-east-1\"\n",
    "            )\n",
    "        )\n",
    "    print(\"Done!\")\n",
    "    return pc\n",
    "\n",
    "\n",
    "def upsert_vectors(\n",
    "        index: Pinecone,\n",
    "        embeddings: np.ndarray,\n",
    "        dataset: dict,\n",
    "        text_field: str = 'highlights',\n",
    "        batch_size: int = 128,\n",
    ") -> Pinecone:\n",
    "    \"\"\"\n",
    "    Upsert vectors to a pinecone index\n",
    "    Args:\n",
    "        index: The pinecone index object\n",
    "        embeddings: The embeddings to upsert\n",
    "        dataset: The dataset containing the metadata\n",
    "        batch_size: The batch size to use for upserting\n",
    "        text_field: The field to embed\n",
    "    Returns:\n",
    "        An updated pinecone index\n",
    "    \"\"\"       \n",
    "    print(\"Upserting the embeddings to the Pinecone index...\")\n",
    "    shape = embeddings.shape\n",
    "    \n",
    "    ids = [str(i) for i in range(shape[0])]\n",
    "    meta = [{text_field: text} for text in dataset[text_field]]\n",
    "    \n",
    "    # create list of (id, vector, metadata) tuples to be upserted\n",
    "    to_upsert = list(zip(ids, embeddings, meta))\n",
    "\n",
    "    for i in tqdm(range(0, shape[0], batch_size)):\n",
    "        i_end = min(i + batch_size, shape[0])\n",
    "        index.upsert(vectors=to_upsert[i:i_end])\n",
    "    return index\n",
    "\n",
    "\n",
    "def augment_prompt(\n",
    "        query: str,\n",
    "        index: Pinecone,\n",
    "        model: SentenceTransformer = SentenceTransformer('all-MiniLM-L6-v2'),\n",
    "        top_k: int = 3,\n",
    ") -> str:\n",
    "    \"\"\"\n",
    "    Augment the prompt with the top 3 results from the knowledge base\n",
    "    Args:\n",
    "        query: The query to augment\n",
    "        index: The vectorstore object\n",
    "        model: The sentence embedding model\n",
    "        top_k: Number of top results to retrieve from the knowlege base\n",
    "    Returns:\n",
    "        str: The augmented prompt\n",
    "    \"\"\"\n",
    "    results = [float(val) for val in list(model.encode(query))]\n",
    "    \n",
    "    # get top k results from knowledge base\n",
    "    query_results = index.query(\n",
    "        vector=results,\n",
    "        top_k=top_k,\n",
    "        include_values=True,\n",
    "        include_metadata=True\n",
    "    )['matches']\n",
    "    text_matches = [match['metadata']['text'] for match in query_results]\n",
    "    \n",
    "    # get the text from the results\n",
    "    source_knowledge = \"\\n\\n\".join(text_matches)\n",
    "    \n",
    "    # feed into an augmented prompt\n",
    "    augmented_prompt = f\"\"\"Using the contexts below, answer as short as you can to the query.\n",
    "    Contexts:\n",
    "    {source_knowledge}\n",
    "    If the answer is not included in the source knowledge - mention it in the answer, don't use the contexts and answer on basis of your knowledge.\n",
    "    Query: {query}\"\"\"\n",
    "    return augmented_prompt, source_knowledge\n",
    "\n",
    "\n",
    "# Additional function for preprocessing and chunking\n",
    "\n",
    "def bioasq_preprocess(dataset, text_field=\"text\"):\n",
    "    \"\"\"Remove the answer at the beginning of the document (specific for the BioASQ dataset\"\"\"\n",
    "    import re\n",
    "    dataset[text_field] = re.sub(r'<answer>.*?<context> ', '', dataset[text_field])\n",
    "    return dataset\n",
    "\n",
    "\n",
    "def chunk_text(\n",
    "        text: str, \n",
    "        window: int = 2, \n",
    "        overlap: int = 0,\n",
    ") -> list:\n",
    "    \"\"\"\n",
    "    Split text into chunks of window size of sentences, with possible overlaping.\n",
    "    Args:\n",
    "        text: input text to split\n",
    "        window: number of consequent sentences in a chunk\n",
    "        overlap: number of overlaping sentences between chunks (in case of window > 1)\n",
    "        preprocess_func: function for text preprocessing\n",
    "    Returns:\n",
    "        list: a list of chunks\n",
    "    \"\"\"\n",
    "    import re\n",
    "    \n",
    "    overlap = overlap if window > overlap else window - 1\n",
    "    sentences = re.split(r'(?<=[.?!])\\s+', text)\n",
    "\n",
    "    chunks = []\n",
    "    for i in range(0, len(sentences), window - overlap):\n",
    "        chunks.append(\" \".join(sentences[i: i + window]))\n",
    "\n",
    "    return chunks\n",
    "\n",
    "\n",
    "def get_semantic_chunks(\n",
    "        chunks: list, \n",
    "        model: SentenceTransformer = SentenceTransformer('all-MiniLM-L6-v2'),\n",
    "        percentile_threshold: int = 90,\n",
    ") -> list:\n",
    "    \"\"\"\n",
    "    Semantic chunking, given a list of sentences or sentences groups (each group will be embedded as a single sentence).\n",
    "    Args:\n",
    "        chunks: list of sentences or sentences groups/chunks\n",
    "        model: the sentence embedding model\n",
    "        percentile_threshold: all sentences distances that above the percentile will be considered as breakpoints to chunk between\n",
    "    Returns:\n",
    "        list: list of semantic chunks\n",
    "    \"\"\"\n",
    "    import numpy as np\n",
    "    import re\n",
    "    from sklearn.metrics.pairwise import cosine_similarity\n",
    "\n",
    "    # embed and calculate cosine distances (1 - cos_sim) between adjacent chunks\n",
    "    embeddings = model.encode(chunks)\n",
    "    distances = np.array([1 - cosine_similarity([embeddings[i]],[embeddings[i + 1]]) for i in range(len(embeddings) - 1)]).reshape(len(embeddings) - 1,)\n",
    "    # find breakpoints (aka divide sentences into groups), this done by percentile value of distances - all above the percentile will be chunked\n",
    "    try:\n",
    "        distance_threshold = np.percentile(distances, percentile_threshold)\n",
    "        breakpoint_indices = [i for i, x in enumerate(distances) if x > distance_threshold]\n",
    "\n",
    "        start_idx = 0\n",
    "        semantic_chunks = []\n",
    "    \n",
    "        for break_idx in breakpoint_indices:\n",
    "            # in case of several sentences in the chunk, there may be overlaping, so need to remove duplicates before chunking\n",
    "            sentences = sum([re.split(r'(?<=[.?!])\\s+', chunk) for chunk in chunks[start_idx: break_idx + 1]], [])\n",
    "            chunk = []\n",
    "            for s in sentences:\n",
    "                if s not in chunk:\n",
    "                    chunk.append(s)\n",
    "            chunk = \" \".join(chunk)\n",
    "            semantic_chunks.append(chunk)\n",
    "            start_idx = break_idx + 1\n",
    "\n",
    "\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "        print(chunks)\n",
    "        print(distances)\n",
    "    \n",
    "    return semantic_chunks\n",
    "\n",
    "\n",
    "def chunk_row(\n",
    "            row, \n",
    "            model: SentenceTransformer = SentenceTransformer('all-MiniLM-L6-v2'),\n",
    "            text_field='text', \n",
    "            window=2, \n",
    "            overlap=1,\n",
    "            percentile_threshold: int = 90,\n",
    "    ) -> list:\n",
    "    \"\"\"\n",
    "    Semantic chunk of text field in a single hugging face dataset row.\n",
    "    Args:\n",
    "        row: a Dataset row (hugging face dataset)\n",
    "        text_field: a text column to chunk\n",
    "        window: number of consequent sentences in a chunk\n",
    "        overlap: number of overlaping sentences between chunks (in case of window > 1)\n",
    "    \"\"\"\n",
    "    chunks = get_semantic_chunks(chunk_text(row[text_field], window=window, overlap=overlap), \n",
    "                                 model, percentile_threshold=percentile_threshold)\n",
    "    result = []\n",
    "    for chunk in chunks:\n",
    "        new_row = row.copy()\n",
    "        new_row[text_field] = chunk\n",
    "        result.append(new_row)\n",
    "    return result\n",
    "\n",
    "\n",
    "def chunk_dataset(\n",
    "        dataset, \n",
    "        model: SentenceTransformer = SentenceTransformer('all-MiniLM-L6-v2'),\n",
    "        text_field: str = 'text',\n",
    "        rec_num: int = 400,\n",
    "        window: int = 2,\n",
    "        overlap: int = 1,\n",
    "        percentile_threshold: int = 90,\n",
    "    ):\n",
    "    \"\"\"\n",
    "    Semantic chunk of text field in the dataset and combining into new dataset, where each chunk will be in a new row, with duplicated data in other cells.\n",
    "    Args:\n",
    "        dataset: the hugging face dataset\n",
    "        model: model for sentence embedding\n",
    "        text_field: the field with text to chunk\n",
    "        window: number of consequent sentences in a chunk\n",
    "        overlap: number of overlaping sentences between chunks (in case of window > 1)\n",
    "    Returns:\n",
    "        chunked_dataset\n",
    "        new_rec_num: updated records number, according to the number of chunks created from the first rec_num rows\n",
    "    \"\"\"\n",
    "    new_rows = []\n",
    "    new_rec_num = 0\n",
    "    for i, row in enumerate(dataset):\n",
    "        chunks = chunk_row(row, model, window=window, overlap=overlap, percentile_threshold=percentile_threshold)\n",
    "        if i < rec_num:\n",
    "            new_rec_num += len(chunks)\n",
    "        new_rows.extend(chunks)\n",
    "\n",
    "    chunked_dataset = datasets.Dataset.from_pandas(pd.DataFrame(new_rows))\n",
    "    return chunked_dataset, new_rec_num\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b9acb6e-d0bd-42cf-aec8-fe0e57e6d53a",
   "metadata": {},
   "source": [
    "## The RAG pipeline\n",
    "Based on tutorial 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad0cbfc7-307b-4b01-b755-9d66c88e72f4",
   "metadata": {},
   "source": [
    "### Loading, embedding and indexing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a75a6b74-0997-4fad-af15-e19888fed6da",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"cohere_key.txt\") as f:\n",
    "    COHERE_API_KEY = f.read().strip()\n",
    "with open(\"pinecone_key.txt\") as f:\n",
    "    PINECONE_API_KEY = f.read().strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "28444d20-c9a7-4f4f-9b4d-2806630e9449",
   "metadata": {},
   "outputs": [],
   "source": [
    "EMBEDDING_MODEL = 'Alibaba-NLP/gte-base-en-v1.5' # best up-to ~0.5gb size model according to huggingface leaderboard (31 place of all models)\n",
    "model = SentenceTransformer(EMBEDDING_MODEL, trust_remote_code=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0b50b914-a081-4cfc-beb5-762f248145e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading and embedding the dataset\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "# Document reading, processing, semantic chinking and embedding\n",
    "DATASET_NAME = 'kroshan/BioASQ'\n",
    "\n",
    "dataset, embeddings = load_and_embedd_dataset(\n",
    "    dataset_name=DATASET_NAME,\n",
    "    rec_num=1000,\n",
    "    model=model,\n",
    "    text_field='text',\n",
    "    shuffle=True,\n",
    "    semantic_chunk=True,\n",
    "    chunk_percentile_threshold=95,\n",
    "    preprocess_func=bioasq_preprocess,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b8420bde-0107-44aa-b2c6-f47ff423d2d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The embeddings shape: (1007, 768)\n"
     ]
    }
   ],
   "source": [
    "print(f\"The embeddings shape: {embeddings.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3c828664-7e91-4ead-a3ca-8649097a9307",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating a Pinecone index...\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "# Create the vector database. We are passing the index_name and the size of our embeddings\n",
    "INDEX_NAME = 'bioasq'\n",
    "pc = create_pinecone_index(INDEX_NAME, embeddings.shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6676cba4-c57d-47db-a338-3c8870fc2d3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Upserting the embeddings to the Pinecone index...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████| 8/8 [00:06<00:00,  1.22it/s]\n"
     ]
    }
   ],
   "source": [
    "# Upsert the embeddings to the Pinecone index\n",
    "index = pc.Index(INDEX_NAME)\n",
    "index_upserted = upsert_vectors(index, embeddings, dataset, text_field='text')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a1e7af79-1d6b-4188-bd42-799a272b7218",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'dimension': 768,\n",
       " 'index_fullness': 0.0,\n",
       " 'namespaces': {'': {'vector_count': 1007}},\n",
       " 'total_vector_count': 1007}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "index.describe_index_stats()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af30cbc0-9328-406d-af7c-6a9a3fe4642d",
   "metadata": {},
   "source": [
    "### Retrieval and augmentation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ba7a77ea-ee2f-4b93-b977-57e34c96fbe1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>question</th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Which molecule is targeted by the drug Gevokizumab?</td>\n",
       "      <td>Effects of gevokizumab on glycemia and inflammatory markers in type 2 diabetes. OBJECTIVE: Metabolic activation of the innate immune system governed by interleukin (IL)-1β contributes to β-cell failure in type 2 diabetes. Gevokizumab is a novel, human-engineered monoclonal anti-IL-1β antibody. We evaluated the safety and biological activity of gevokizumab in patients with type 2 diabetes. RESEARCH DESIGN AND METHODS: In a placebo-controlled, dose-escalation study, a total of 98 patients were randomly assigned to placebo (17 subjects) or gevokizumab (81 subjects) at increasing doses and dosing schedules. The primary objective of the study was to evaluate the safety profile of gevokizumab in type 2 diabetes. The secondary objectives were to assess pharmacokinetics for different dose levels, routes of administration, and regimens and to assess biological activity. RESULTS: The study drug was well tolerated with no serious adverse events. There was one hypoglycemic event whereupon concomitant insulin treatment had to be reduced. Clearance of gevokizumab was consistent with that for a human IgG(2), with a half-life of 22 days. In the combined intermediate-dose group (single doses of 0.03 and 0.1 mg/kg), the mean placebo-corrected decrease in glycated hemoglobin was 0.11, 0.44, and 0.85% after 1, 2 (P = 0.017), and 3 (P = 0.049) months, respectively, along with enhanced C-peptide secretion, increased insulin sensitivity, and a reduction in C-reactive protein and spontaneous and inducible cytokines. CONCLUSIONS: This novel IL-1β-neutralizing antibody improved glycemia, possibly via restored insulin production and action, and reduced inflammation in patients with type 2 diabetes. This therapeutic agent may be able to be used on a once-every-month or longer schedule.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Which clotting factor is inhibited by betrixaban?</td>\n",
       "      <td>Oral and parenteral anticoagulants: new kids on the block. Well-documented drawbacks of traditional anticoagulants have lead to the quest for an ideal anticoagulant resulting in a surge of novel anticoagulant molecules.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Against which protein is the antibody used for immonostaining of Lewy bodies raised?</td>\n",
       "      <td>Nigral and cortical Lewy bodies and dystrophic nigral neurites in Parkinson's disease and cortical Lewy body disease contain alpha-synuclein immunoreactivity. A mutation in the alpha-synuclein gene has recently been linked to some cases of familial Parkinson's disease (PD). We characterized the expression of this presynaptic protein in the midbrain, striatum, and temporal cortex of control, PD, and dementia with Lewy bodies (DLB) brain. Control brain showed punctate pericellular immunostaining.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>What tyrosine kinase, involved in a Philadelphia- chromosome positive chronic myelogenous leukemia, is the target of Imatinib (Gleevec)?</td>\n",
       "      <td>Drug responses of imatinib mesylate-resistant cells: synergism of imatinib with other chemotherapeutic drugs. Imatinib mesylate (STI571, Glivec, Gleevec) is a powerful inhibitor of the tyrosine kinase activity of Bcr-Abl, the oncoprotein responsible for chronic myeloid leukemia (CML). The drug shows great efficacy in chronic phase, but is less effective in maintaining hematologic remissions in blast crisis patients. Our group has previously described several cell lines made resistant to imatinib. We now examine the question of cross-resistance to other chemotherapeutic drugs used in CML. Four paired imatinib-sensitive/resistant CML cell lines were assessed by caspase-3 and MTS assays for their proliferative response to cytosine arabinoside (Ara-C), daunorubicin (DNR), homoharringtonine (HHT) and hydroxyurea (HU), either alone or in combination with imatinib. Primary blasts from advanced-stage CML patients refractory to imatinib therapy were studied by semi-solid media clonogenic assays. We found that these drugs are generally capable of major inhibition of proliferation of the CML cell lines, although differential responses to DNR and HHT were noted between some sensitive and resistant cell line pairs, implying that resistance to imatinib may confer a growth advantage under such conditions. The four drugs were also effective in preventing the formation of progenitor cell colonies from CML patients both before treatment with imatinib, and after relapse on the drug. Isobolographic analysis implied that these drugs will generally combine well with imatinib, and in some cases will be synergistic.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>The drug JTV519 is derivative of which group of chemical compounds?</td>\n",
       "      <td>JTV-519, a novel cardioprotective agent, improves the contractile recovery after ischaemia-reperfusion in coronary perfused guinea-pig ventricular muscles. A newly synthesized benzothiazepine derivative, JTV-519 (JT) has been reported to be cardioprotective. However, the precise mechanism underlying the cardioprotective effect of this drug is unknown.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Mutation of which gene is associated with McLeod syndrome?</td>\n",
       "      <td>McLeod syndrome: life-long neuropsychiatric disorder due to a novel mutation of the XK gene. A 50-year-old man presented with worsening, virtually lifelong, chorea and progressive behavioural disturbance, involving disinhibition and hoarding, over 10 years. Clinical assessment revealed chorea, dysarthria, areflexia, an inappropriately jovial, impulsive manner and neuropsychological evidence of frontosubcortical dysfunction. Investigation results included an elevated creatine kinase, caudate atrophy and hypoperfusion, acanthocytes in the peripheral blood and the McLeod phenotype.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>What is the synonym of the lubag disease?</td>\n",
       "      <td>Rare causes of dystonia parkinsonism. The list of genetic causes of syndromes of dystonia parkinsonism grows constantly. As a consequence, the diagnosis becomes more and more challenging for the clinician. Here, we summarize the important causes of dystonia parkinsonism including autosomal-dominant, recessive, and x-linked forms. We cover dopa-responsive dystonia, Wilson's disease, Parkin-, PINK1-, and DJ-1-associated parkinsonism (PARK2, 6, and 7), x-linked dystonia-parkinsonism/Lubag (DYT3), rapid-onset dystonia-parkinsonism (DYT12) and DYT16 dystonia, the syndromes of Neurodegeneration with Brain Iron Accumulation (NBIA) including pantothenate kinase (PANK2)- and PLA2G6 (PARK14)-associated neurodegeneration, neuroferritinopathy, Kufor-Rakeb disease (PARK9) and the recently described SENDA syndrome; FBXO7-associated neurodegeneration (PARK15), autosomal-recessive spastic paraplegia with a thin corpus callosum (SPG11), and dystonia parkinsonism due to mutations in the SLC6A3 gene encoding the dopamine transporter. They have in common that in all these syndromes there may be a combination of dystonic and parkinsonian features, which may be complicated by pyramidal tract involvement. The aim of this review is to familiarize the clinician with the phenotypes of these disorders.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Which is the transcript responsible for X-chromosome inactivation?</td>\n",
       "      <td>X-inactivation: quantitative predictions of protein interactions in the Xist network. The transcriptional silencing of one of the female X-chromosomes is a finely regulated process that requires accumulation in cis of the long non-coding RNA X-inactive-specific transcript (Xist) followed by a series of epigenetic modifications. Little is known about the molecular machinery regulating initiation and maintenance of chromosomal silencing.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>What is the name of Bruton's tyrosine kinase inhibitor that can be used for treatment of chronic lymphocytic leukemia?</td>\n",
       "      <td>Targeting the microenvironment in chronic lymphocytic leukemia is changing the therapeutic landscape. PURPOSE OF REVIEW: Despite ongoing efforts to decipher the cancer genome, discoveries of new targetable genetic lesions within cancer cells are rare. Therefore, alternative approaches are needed. Signals from the microenvironment are increasingly recognized as drivers of disease progression in hematologic and solid cancers. Consequently, there is growing interest in targeting the tumor-microenvironment cross-talk. This review highlights recent therapeutic advances in targeting the microenvironment in chronic lymphocytic leukemia (CLL). RECENT FINDINGS: CLL is the poster child for microenvironment-dependent malignancies, because the clonal CLL B cells are highly dependent on external signals for maintenance and expansion. These pathways recapitulate those responsible for normal B-cell expansion in germinal centers. The most prominent, conserved mechanism is B-cell receptor (BCR) signaling, which promotes CLL cell survival and expansion in lymphatic tissue areas designated proliferation centers. BCR signaling now can be targeted by new targeted kinase inhibitors. SUMMARY: Small molecule inhibitors of BCR signaling kinases, Bruton's tyrosine kinase (Btk) inhibitor ibrutinib and the phosphoinositide 3'-kinase delta (PI3Kδ) inhibitor GS-1101, are currently transforming the landscape of CLL therapy. This development exemplifies that the microenvironment has become a lively successful area of translational research.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Which is the process that Conserved noncoding elements mostly regulate?</td>\n",
       "      <td>Disruption of long-distance highly conserved noncoding elements in neurocristopathies. One of the key discoveries of vertebrate genome sequencing projects has been the identification of highly conserved noncoding elements (CNEs). Some characteristics of CNEs include their high frequency in mammalian genomes, their potential regulatory role in gene expression, and their enrichment in gene deserts nearby master developmental genes.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                   question  \\\n",
       "0                                                                                       Which molecule is targeted by the drug Gevokizumab?   \n",
       "1                                                                                         Which clotting factor is inhibited by betrixaban?   \n",
       "2                                                      Against which protein is the antibody used for immonostaining of Lewy bodies raised?   \n",
       "3  What tyrosine kinase, involved in a Philadelphia- chromosome positive chronic myelogenous leukemia, is the target of Imatinib (Gleevec)?   \n",
       "4                                                                       The drug JTV519 is derivative of which group of chemical compounds?   \n",
       "5                                                                                Mutation of which gene is associated with McLeod syndrome?   \n",
       "6                                                                                                 What is the synonym of the lubag disease?   \n",
       "7                                                                        Which is the transcript responsible for X-chromosome inactivation?   \n",
       "8                    What is the name of Bruton's tyrosine kinase inhibitor that can be used for treatment of chronic lymphocytic leukemia?   \n",
       "9                                                                   Which is the process that Conserved noncoding elements mostly regulate?   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           text  \n",
       "0  Effects of gevokizumab on glycemia and inflammatory markers in type 2 diabetes. OBJECTIVE: Metabolic activation of the innate immune system governed by interleukin (IL)-1β contributes to β-cell failure in type 2 diabetes. Gevokizumab is a novel, human-engineered monoclonal anti-IL-1β antibody. We evaluated the safety and biological activity of gevokizumab in patients with type 2 diabetes. RESEARCH DESIGN AND METHODS: In a placebo-controlled, dose-escalation study, a total of 98 patients were randomly assigned to placebo (17 subjects) or gevokizumab (81 subjects) at increasing doses and dosing schedules. The primary objective of the study was to evaluate the safety profile of gevokizumab in type 2 diabetes. The secondary objectives were to assess pharmacokinetics for different dose levels, routes of administration, and regimens and to assess biological activity. RESULTS: The study drug was well tolerated with no serious adverse events. There was one hypoglycemic event whereupon concomitant insulin treatment had to be reduced. Clearance of gevokizumab was consistent with that for a human IgG(2), with a half-life of 22 days. In the combined intermediate-dose group (single doses of 0.03 and 0.1 mg/kg), the mean placebo-corrected decrease in glycated hemoglobin was 0.11, 0.44, and 0.85% after 1, 2 (P = 0.017), and 3 (P = 0.049) months, respectively, along with enhanced C-peptide secretion, increased insulin sensitivity, and a reduction in C-reactive protein and spontaneous and inducible cytokines. CONCLUSIONS: This novel IL-1β-neutralizing antibody improved glycemia, possibly via restored insulin production and action, and reduced inflammation in patients with type 2 diabetes. This therapeutic agent may be able to be used on a once-every-month or longer schedule.  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Oral and parenteral anticoagulants: new kids on the block. Well-documented drawbacks of traditional anticoagulants have lead to the quest for an ideal anticoagulant resulting in a surge of novel anticoagulant molecules.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           Nigral and cortical Lewy bodies and dystrophic nigral neurites in Parkinson's disease and cortical Lewy body disease contain alpha-synuclein immunoreactivity. A mutation in the alpha-synuclein gene has recently been linked to some cases of familial Parkinson's disease (PD). We characterized the expression of this presynaptic protein in the midbrain, striatum, and temporal cortex of control, PD, and dementia with Lewy bodies (DLB) brain. Control brain showed punctate pericellular immunostaining.  \n",
       "3                                                                                                                                                                           Drug responses of imatinib mesylate-resistant cells: synergism of imatinib with other chemotherapeutic drugs. Imatinib mesylate (STI571, Glivec, Gleevec) is a powerful inhibitor of the tyrosine kinase activity of Bcr-Abl, the oncoprotein responsible for chronic myeloid leukemia (CML). The drug shows great efficacy in chronic phase, but is less effective in maintaining hematologic remissions in blast crisis patients. Our group has previously described several cell lines made resistant to imatinib. We now examine the question of cross-resistance to other chemotherapeutic drugs used in CML. Four paired imatinib-sensitive/resistant CML cell lines were assessed by caspase-3 and MTS assays for their proliferative response to cytosine arabinoside (Ara-C), daunorubicin (DNR), homoharringtonine (HHT) and hydroxyurea (HU), either alone or in combination with imatinib. Primary blasts from advanced-stage CML patients refractory to imatinib therapy were studied by semi-solid media clonogenic assays. We found that these drugs are generally capable of major inhibition of proliferation of the CML cell lines, although differential responses to DNR and HHT were noted between some sensitive and resistant cell line pairs, implying that resistance to imatinib may confer a growth advantage under such conditions. The four drugs were also effective in preventing the formation of progenitor cell colonies from CML patients both before treatment with imatinib, and after relapse on the drug. Isobolographic analysis implied that these drugs will generally combine well with imatinib, and in some cases will be synergistic.  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             JTV-519, a novel cardioprotective agent, improves the contractile recovery after ischaemia-reperfusion in coronary perfused guinea-pig ventricular muscles. A newly synthesized benzothiazepine derivative, JTV-519 (JT) has been reported to be cardioprotective. However, the precise mechanism underlying the cardioprotective effect of this drug is unknown.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     McLeod syndrome: life-long neuropsychiatric disorder due to a novel mutation of the XK gene. A 50-year-old man presented with worsening, virtually lifelong, chorea and progressive behavioural disturbance, involving disinhibition and hoarding, over 10 years. Clinical assessment revealed chorea, dysarthria, areflexia, an inappropriately jovial, impulsive manner and neuropsychological evidence of frontosubcortical dysfunction. Investigation results included an elevated creatine kinase, caudate atrophy and hypoperfusion, acanthocytes in the peripheral blood and the McLeod phenotype.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              Rare causes of dystonia parkinsonism. The list of genetic causes of syndromes of dystonia parkinsonism grows constantly. As a consequence, the diagnosis becomes more and more challenging for the clinician. Here, we summarize the important causes of dystonia parkinsonism including autosomal-dominant, recessive, and x-linked forms. We cover dopa-responsive dystonia, Wilson's disease, Parkin-, PINK1-, and DJ-1-associated parkinsonism (PARK2, 6, and 7), x-linked dystonia-parkinsonism/Lubag (DYT3), rapid-onset dystonia-parkinsonism (DYT12) and DYT16 dystonia, the syndromes of Neurodegeneration with Brain Iron Accumulation (NBIA) including pantothenate kinase (PANK2)- and PLA2G6 (PARK14)-associated neurodegeneration, neuroferritinopathy, Kufor-Rakeb disease (PARK9) and the recently described SENDA syndrome; FBXO7-associated neurodegeneration (PARK15), autosomal-recessive spastic paraplegia with a thin corpus callosum (SPG11), and dystonia parkinsonism due to mutations in the SLC6A3 gene encoding the dopamine transporter. They have in common that in all these syndromes there may be a combination of dystonic and parkinsonian features, which may be complicated by pyramidal tract involvement. The aim of this review is to familiarize the clinician with the phenotypes of these disorders.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       X-inactivation: quantitative predictions of protein interactions in the Xist network. The transcriptional silencing of one of the female X-chromosomes is a finely regulated process that requires accumulation in cis of the long non-coding RNA X-inactive-specific transcript (Xist) followed by a series of epigenetic modifications. Little is known about the molecular machinery regulating initiation and maintenance of chromosomal silencing.  \n",
       "8                                                                                                                                                                                                                                                                Targeting the microenvironment in chronic lymphocytic leukemia is changing the therapeutic landscape. PURPOSE OF REVIEW: Despite ongoing efforts to decipher the cancer genome, discoveries of new targetable genetic lesions within cancer cells are rare. Therefore, alternative approaches are needed. Signals from the microenvironment are increasingly recognized as drivers of disease progression in hematologic and solid cancers. Consequently, there is growing interest in targeting the tumor-microenvironment cross-talk. This review highlights recent therapeutic advances in targeting the microenvironment in chronic lymphocytic leukemia (CLL). RECENT FINDINGS: CLL is the poster child for microenvironment-dependent malignancies, because the clonal CLL B cells are highly dependent on external signals for maintenance and expansion. These pathways recapitulate those responsible for normal B-cell expansion in germinal centers. The most prominent, conserved mechanism is B-cell receptor (BCR) signaling, which promotes CLL cell survival and expansion in lymphatic tissue areas designated proliferation centers. BCR signaling now can be targeted by new targeted kinase inhibitors. SUMMARY: Small molecule inhibitors of BCR signaling kinases, Bruton's tyrosine kinase (Btk) inhibitor ibrutinib and the phosphoinositide 3'-kinase delta (PI3Kδ) inhibitor GS-1101, are currently transforming the landscape of CLL therapy. This development exemplifies that the microenvironment has become a lively successful area of translational research.  \n",
       "9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             Disruption of long-distance highly conserved noncoding elements in neurocristopathies. One of the key discoveries of vertebrate genome sequencing projects has been the identification of highly conserved noncoding elements (CNEs). Some characteristics of CNEs include their high frequency in mammalian genomes, their potential regulatory role in gene expression, and their enrichment in gene deserts nearby master developmental genes.  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_colwidth', None)\n",
    "pd.set_option('display.width', None)\n",
    "pd.set_option('display.max_rows', None)\n",
    "\n",
    "pd_dataset = dataset.to_pandas()\n",
    "pd_dataset.head(10)  # show some random records (the dataset was shuffled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "32b0f4fc-a107-4b64-bcc0-d1e89f8394f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1. The drug JTV519 is derivative of which group of chemical compounds?\n",
      "Model's answer: The drug JTV519 is a derivative of the chemical compound group called benzopyran. Specifically, it is a member of the chroman family, which is a type of benzopyran. Benzopyrans are characterized by a fused ring structure consisting of a six-membered benzene ring and a pyran ring, which is a five-membered ring containing four carbon atoms and one oxygen atom.\n",
      "\n",
      "In the case of JTV519, the benzopyran structure is modified to include specific substituents that impart its pharmacological properties. These modifications are designed to target specific biological mechanisms and achieve the desired therapeutic effects.\n",
      "\n",
      "JTV519, also known as k252a, is a potent and selective inhibitor of certain types of calcium channels, particularly the L-type calcium channels. It has been investigated for its potential therapeutic benefits in various cardiovascular and neurological disorders. The development and study of JTV519 have contributed to our understanding of calcium channel function and the potential treatment options for conditions associated with calcium channel dysfunction.\n",
      "\n",
      "Correct answer: benzothiazepine\n",
      "\n",
      "2. What is the synonym of the lubag disease?\n",
      "Model's answer: The synonym of the \"lubag disease\" is:\n",
      "\n",
      "**Lou Gehrig's Disease** - This term is widely used and recognized as a synonym for amyotrophic lateral sclerosis (ALS). It is named after the famous baseball player, Lou Gehrig, who was diagnosed with this disease and brought significant attention to it.\n",
      "\n",
      "Correct answer: x-linked dystonia-parkinsonism\n",
      "\n",
      "3. Which medication should be administered when managing patients with suspected acute opioid overdose?\n",
      "Model's answer: When managing patients with a suspected acute opioid overdose, the medication that should be administered is naloxone. Naloxone is an opioid antagonist that competes with opioids for binding sites in the brain, reversing the life-threatening effects of opioid overdose. It works by blocking the opioid receptors, counteracting the effects of opioids, and restoring respiratory function.\n",
      "\n",
      "Naloxone can be administered through various routes, including intravenous (IV), intramuscular (IM), subcutaneous, intranasal, and even through a nebulizer in some cases. The route of administration may depend on the availability of resources, the patient's condition, and the training of the healthcare provider.\n",
      "\n",
      "It is important to note that naloxone has a shorter duration of action compared to many opioids. Therefore, continuous monitoring of the patient is crucial after administering naloxone, as the effects of opioids can outlast those of naloxone, potentially leading to a recurrence of respiratory depression and other overdose symptoms.\n",
      "\n",
      "In addition to administering naloxone, ensuring adequate airway management, providing supportive care, and monitoring vital signs are essential components of managing patients with suspected acute opioid overdose. If you are a bystander or caregiver and suspect an opioid overdose, it is always best to call for emergency medical assistance immediately, as prompt treatment can save lives.\n",
      "\n",
      "Correct answer: naloxone\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Anecdote examples of queries with wrong answers\n",
    "queries = [\"The drug JTV519 is derivative of which group of chemical compounds?\",\n",
    "           \"What is the synonym of the lubag disease?\",\n",
    "           \"Which medication should be administered when managing patients with suspected acute opioid overdose?\"\n",
    "          ]\n",
    "answers = [\"benzothiazepine\",\n",
    "           \"x-linked dystonia-parkinsonism\",\n",
    "           \"naloxone\"\n",
    "          ]\n",
    "\n",
    "co = cohere.Client(api_key=COHERE_API_KEY)\n",
    "\n",
    "for i in range(len(queries)):\n",
    "    response = co.chat(\n",
    "            model='command-r-plus',\n",
    "            message=queries[i],\n",
    "        )\n",
    "    print(f\"{i+1}. {queries[i]}\")\n",
    "    print(f\"Model's answer: {response.text}\")\n",
    "    print(f\"\\nCorrect answer: {answers[i]}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8a359797-55cf-4dea-b5e0-5d4020df246e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1. The drug JTV519 is derivative of which group of chemical compounds?\n",
      "Model's answer: JTV519 is a derivative of 1,4-benzothiazepine.\n",
      "\n",
      "Correct answer: benzothiazepine\n",
      "\n",
      "2. What is the synonym of the lubag disease?\n",
      "Model's answer: X-linked dystonia-parkinsonism.\n",
      "\n",
      "Correct answer: x-linked dystonia-parkinsonism\n",
      "\n",
      "3. Which medication should be administered when managing patients with suspected acute opioid overdose?\n",
      "Model's answer: Naloxone.\n",
      "\n",
      "Correct answer: naloxone\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Anecdote examples of queries with wrong answers, now with augmented prompts\n",
    "queries = [\"The drug JTV519 is derivative of which group of chemical compounds?\",\n",
    "           \"What is the synonym of the lubag disease?\",\n",
    "           \"Which medication should be administered when managing patients with suspected acute opioid overdose?\"\n",
    "          ]\n",
    "answers = [\"benzothiazepine\",\n",
    "           \"x-linked dystonia-parkinsonism\",\n",
    "           \"naloxone\"\n",
    "          ]\n",
    "for i in range(len(queries)):\n",
    "    augmented_prompt, source_knowledge = augment_prompt(queries[i], model=model, index=index)\n",
    "    response = co.chat(\n",
    "            model='command-r-plus',\n",
    "            message=augmented_prompt,\n",
    "        )\n",
    "    print(f\"{i+1}. {queries[i]}\")\n",
    "    print(f\"Model's answer: {response.text}\")\n",
    "    print(f\"\\nCorrect answer: {answers[i]}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python daav",
   "language": "python",
   "name": "azureml_py38_pt_and_tf"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
